aws-datastore/src/main/java/com/amplifyframework/datastore/syncengine/RequestRetry.kt

```diff
@@ -0,0 +1,45 @@
+package com.amplifyframework.datastore.syncengine
+
+import com.amplifyframework.AmplifyException
+
+import io.reactivex.rxjava3.core.Single
+import io.reactivex.rxjava3.core.SingleEmitter
+import java.util.concurrent.TimeUnit
+import kotlin.math.pow
+
+
+class RequestRetry(private val maxExponent: Int = 8,
+                    private val jitterFactor: Int = 100,
+                    private val maxAttempts: Int = 3) {
+
+    private var numberOfAttempts = 0
+
+
+        fun <T> retry(single: Single<T>, skipExceptions: List<Class<out Throwable>?>): Single<T> {
+            return Single.create { emitter -> call(single, emitter, 0, skipExceptions) }
```

Comment:
As this class is currently designed, you will need to create an instance of `RequestRetry` for each `Single` that you want to retry.   This is because `numberOfAttempts` is only set to 0 when the class is instantiated, and is never reset after that.  This is not a great design because it might accidentally get used to retry more than one thing (you actually are doing this in the `SyncProcessor` currently).   I think you can solve this by removing the `numberOfAttempts` variable, and just adding a `numAttempts` to the `call(...)` method.  The first invocation will set it to `MAX_ATTEMPTS`, like this:
```suggestion
            return Single.create { emitter -> call(single, emitter, 0, MAX_ATTEMPTS, skipExceptions) }
```

Then, inside the `call` method, when you recursively call it again, decrement `numAttempts` by 1.

Reply:
Great catch, thanks!

Reply:
Done.